### PR TITLE
Add Files, Clipboard, and first Advanced Apps Patterns

### PR DESCRIPTION
**Pull request by tomayac**
Fixes https://github.com/GoogleChrome/web.dev/issues/7439

Changes proposed in this pull request:

- Add Files and Clipboard Patterns

### CODE REVIEW

**Comment by tomayac** [comment]
Hi @devnook,

In https://github.com/GoogleChrome/web.dev/pull/7645/commits/8b7b2f7b433e9d6dbc98dd355de37636e272c6fb, I have now reflected your changes from https://github.com/GoogleChrome/web.dev/pull/7875; and it _almost_ works. Since the service worker lives in `assets/sw.js`, I cannot broaden its `scope` unfortunately:

https://github.com/GoogleChrome/web.dev/blob/8b7b2f7b433e9d6dbc98dd355de37636e272c6fb/src/site/content/en/patterns/advanced-apps/badges/demo.njk#L35-L39

Screen Shot 2022-05-18 at 13 04 34

Any idea? Can we somehow smuggle `sw.js` one folder level up?

File: src/site/content/en/patterns/advanced-apps/badges/index.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+---
+layout: pattern
+title: How to create an app badge
+date: 2022-05-18
+description: |
+  Learn how to create an app badge.
+static:
+  - assets/manifest.json
+  - assets/sw.js
```
Comment (devnook) [inline src/site/content/en/patterns/advanced-apps/badges/index.md:9]: I suggest you just move sw.js up a level to badges/sw.js and update the path here.

**Comment by devnook** [comment]
> Hi @devnook,
> 
> In [8b7b2f7](https://github.com/GoogleChrome/web.dev/commit/8b7b2f7b433e9d6dbc98dd355de37636e272c6fb), I have now reflected your changes from #7875; and it _almost_ works. Since the service worker lives in `assets/sw.js`, I cannot broaden its `scope` unfortunately:
> 
> https://github.com/GoogleChrome/web.dev/blob/8b7b2f7b433e9d6dbc98dd355de37636e272c6fb/src/site/content/en/patterns/advanced-apps/badges/demo.njk#L35-L39
> 
> Screen Shot 2022-05-18 at 13 04 34
> 
> Any idea? Can we somehow smuggle `sw.js` one folder level up?

I suggest you just move sw.js up a level to badges/sw.js and update the path in index.md `static` field. If you still want to have it displayed in the code tabs, then probably you'd need to repeat it in assets/ directory too.

File: src/site/content/en/patterns/clipboard/paste-files/index.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+---
+layout: pattern
+title: How to paste files
+date: 2022-07-18
+authors:
+  - charistheodoulou
```
Comment (tomayac) [inline src/site/content/en/patterns/clipboard/paste-files/index.md:6]:
```suggestion
  - harrytheodoulou
```

File: src/site/content/en/patterns/clipboard/paste-images/index.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
+---
+layout: pattern
+title: How to paste images
+date: 2022-08-01
+authors:
+  - charistheodoulou
```
Comment (tomayac) [inline src/site/content/en/patterns/clipboard/paste-images/index.md:6]:
```suggestion
  - harrytheodoulou
```

File: src/site/content/en/patterns/clipboard/paste-text/index.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
+---
+layout: pattern
+title: How to paste text
+date: 2022-07-28
+authors:
+  - charistheodoulou
```
Comment (tomayac) [inline src/site/content/en/patterns/clipboard/paste-text/index.md:6]:
```suggestion
  - harrytheodoulou
```

File: src/site/content/en/blog/new-patterns-for-amazing-apps/index.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+### Saving files
+
+When it comes to saving, the go-to approach is to use the
+[File System Access API](/file-system-access/), so you end up with a `FileSystemFileHandle`, which
+allows you to implement a true [save, edit, save flow](/excalidraw-and-fugu/#saving-files). The next
+best thing is to fall back to a classic `<a download>`, which likewise lets the user save data, but
+has the downside of creating new files on each download, so they end up with `my-video.mp4`,
+`my-video.mp4 (1)`, `my-video.mp4 (2)`, etc.
```
Comment (beaufortfrancois) [inline src/site/content/en/blog/new-patterns-for-amazing-apps/index.md:62]:
```suggestion
`my-video (1).mp4`, `my-video (2).mp4`, etc.
```

**Review by beaufortfrancois — COMMENTED** [review]
LGTM! You may want to edit all of `date:` to today.

File: src/site/content/en/blog/new-patterns-for-amazing-apps/index.md (hunk 5 above)
Comment (tomayac) [reply src/site/content/en/blog/new-patterns-for-amazing-apps/index.md:62]: Ah, of course… `:-)` Merci!

File: src/site/content/en/patterns/advanced-apps/contacts/index.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+[Contact Picker API](https://developer.mozilla.org/docs/Web/API/Contact_Picker_API), which allows
+users to select entries from their contact list and share limited details of the selected entries
+with your app. Several properties like `name`, `email`, `tel`, `address`, and `icon` are available.
+To find out about the concretely supported properties, call `navigator.contacts.getProperties()`. To
+allow the user to select multiple contacts, pass `{multiple: true}` as the second parameter of
+`navigator.contacts.select()`.
+
+{% BrowserCompat 'api.AudioContext.outputLatency' %}
+
+The Contact Picker API is available in the Android version of Chrome starting in version 80 and up.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/contacts/index.md:30]:
```suggestion
The Contact Picker API is available in the Android version of Chrome from version 80.
```

File: src/site/content/en/patterns/advanced-apps/multiple-screens/index.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+---
+
+## The modern way
+
+### Using `window.getScreenDetails()`
+
+To make sure your browser supports the `getScreenDetails()` method, you need to check if it exists
+on the `window` object. Then, call `window.getScreenDetails()` to get attached screens. Adding an
+event listener to adapt to changed screen details allows you to manage multiple screens according to
+the needs of your app.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/multiple-screens/index.md:21]:
```suggestion
To make sure your browser supports the `getScreenDetails()` method, first check if it exists
on the `window` object. Then, call `window.getScreenDetails()` to get attached screens. Adding an
event listener to adapt to changed screen details allows you to manage multiple screens according to
the needs of your app.
```

File: src/site/content/en/patterns/clipboard/paste-images/index.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+});
+```
+
+{% BrowserCompat 'api.Clipboard' %}
+
+## Progressive enhancement
+
+For browsers that do not support the Async Clipboard API, it is impossible to access the user's clipboard programmatically (i.e., on a button click). Thus for accessing a user's clipboard on a `paste` event, you can use the Async Clipboard API and fall back to the (synchronous) Clipboard API.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/clipboard/paste-images/index.md:61]:
```suggestion
For browsers that do not support the Async Clipboard API, it is impossible to access the user's clipboard programmatically (for example, on a button click). Thus for accessing a user's clipboard on a `paste` event, you can use the Async Clipboard API and fall back to the (synchronous) Clipboard API.
```

File: src/site/content/en/patterns/clipboard/paste-text/index.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+height: 800
+description: >
+  Pasting text into the browser consists of using the readText() method of the Async Clipboard API.
+---
+
+## The modern way
+
+### Using the Async Clipboard API
+
+To read text from the user's clipboard programmatically, i.e., after a button click, you can use the [`readText()`](https://developer.mozilla.org/docs/Web/API/Clipboard/readText) method of the [Async Clipboard API](https://developer.mozilla.org/docs/Web/API/Clipboard_API). If permissions to read clipboard have not been granted yet, the call to `navigator.clipboard.readText()` will request it upon the first call to the method.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/clipboard/paste-text/index.md:16]:
```suggestion
To read text from the user's clipboard programmatically, for example, after a button click, you can use the [`readText()`](https://developer.mozilla.org/docs/Web/API/Clipboard/readText) method of the [Async Clipboard API](https://developer.mozilla.org/docs/Web/API/Clipboard_API). If permissions to read clipboard have not been granted yet, the call to `navigator.clipboard.readText()` will request it upon the first call to the method.
```

File: src/site/content/en/patterns/files/handle-files-opened-from-the-file-explorer/index.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+## The classic way
+
+### Using the classic `DataTransferItem.getAsFile()` method
+
+If the File Handling API is not supported, you can still drag and drop files from the file explorer
+into the app. The `DataTransferItem.getAsFile()` method returns the drag data item's `File` object.
+If the item is not a file, this method returns `null`. While you can read the file, there is no way
+to write back to it. This method has the disadvantage that it does not support directories.
+{% BrowserCompat 'api.DataTransferItem.getAsFile' %}
```
Comment (rachelandrew) [inline src/site/content/en/patterns/files/handle-files-opened-from-the-file-explorer/index.md:65]:
```suggestion

{% BrowserCompat 'api.DataTransferItem.getAsFile' %}
```

File: src/site/content/en/patterns/files/receive-shared-files/index.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+---
+
+## The modern way
+
+### Using the Web Share Target API
+
+First, declare a `share_target` in your web app manifest that lists an `action` (a URL to handle
+shared files at), a `method` (`"POST"` for files) and an `enctype` (`"multipart/form-data"` for
+files) and a `params` object that contains a `files` property with an array of objects with a `name`
+and `accept` property that list the sharable file types and the name to obtain them.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/files/receive-shared-files/index.md:25]:
```suggestion
First, declare a `share_target` in your web app manifest that lists an `action` (your URL that handles
shared files), a `method` (`"POST"` for files) and an `enctype` (`"multipart/form-data"` for
files) and a `params` object that contains a `files` property with an array of objects with a `name`
and `accept` property that list the sharable file types and the name to obtain them.
```

File: src/site/content/en/patterns/advanced-apps/audio-effects/index.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+description: >
+  Learn how to use the Web Audio API to add effects to audio sources.
+height: 800
+---
+
+## The modern way
+The Web Audio API uses the `AudioContext()` interface to manage sources, filters, and destinations. Once you’ve created a new `AudioContext()`, create an audio source node, like an `AudioBufferSourceNode` or `OscillatorNode`.  As an example, consider a basic oscillator with a [low pass filter](https://en.wikipedia.org/wiki/Low-pass_filter) applied.
+
+{% BrowserCompat 'api.AudioContext' %}
+### Using the `createBiquadFilter()` method
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/audio-effects/index.md:16]:
```suggestion

### Using the `createBiquadFilter()` method
```

File: src/site/content/en/patterns/advanced-apps/audio-effects/index.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+```js
+const audioCtx = new (window.AudioContext || window.webkitAudioContext)();
+const oscillator = audioCtx.createOscillator()
+oscillator.type = 'sine';
+ // Value is in hertz.
+oscillator.frequency.setValueAtTime(420, audioCtx.currentTime);
+```
+
+Next, create an effect node using `createBiquadFilter()`. Set the type to `lowpass` and the frequency to start 1 second after the start of playback. Then connect the `biquadFilter` to the `oscillator`.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/audio-effects/index.md:28]:
```suggestion
Next, create an effect node using `createBiquadFilter()`. Set the type to `lowpass` and the frequency to start one second after the start of playback. Then connect the `biquadFilter` to the `oscillator`.
```

File: src/site/content/en/patterns/advanced-apps/audio-effects/index.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+```js
+const biquadFilter = audioCtx.createBiquadFilter();
+biquadFilter.type = 'lowpass';
+biquadFilter.frequency.setValueAtTime(200, audioCtx.currentTime + 1);
+oscillator.connect(biquadFilter);
+```
+
+Finally connect the `biquadFilter` to the destination of the `audioCtx`, before starting the `oscillator` and stopping it after 2 seconds of playback.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/audio-effects/index.md:37]:
```suggestion
Finally connect the `biquadFilter` to the destination of the `audioCtx`, before starting the `oscillator` and stopping it after two seconds of playback.
```

File: src/site/content/en/patterns/advanced-apps/audio-effects/index.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+```js
+biquadFilter.connect(audioCtx.destination);
+oscillator.start();
+oscillator.stop(2);
+```
+
+The sound from the oscillator is passed through the unaffected filter, and on to the destination, which is your computer's speakers. After 1 second of playback, the `lowpass` filter comes into effect. After 2 seconds, the oscillator is stopped.
```
Comment (rachelandrew) [inline src/site/content/en/patterns/advanced-apps/audio-effects/index.md:45]:
```suggestion
The sound from the oscillator is passed through the unaffected filter, and on to the destination, which is your computer's speakers. After one second of playback, the `lowpass` filter comes into effect. After two seconds, the oscillator is stopped.
```

**Review by rachelandrew — APPROVED** [review]
Approved with some very small styleguide nits.

**Comment by tomayac** [comment]
@devnook Could you flip the eng LGTM approval bit for [`src/lib/patterns.js`](https://github.com/GoogleChrome/web.dev/pull/7645/files#diff-43e1f65fe4e39c60b36b3c98cac54aa1de71cc5ca9a77c2876dde4f7e2b76171), which makes sure the "CodePen" doesn't show images?

**Review by devnook — APPROVED** [review]

**Pull request merged**